sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala

```diff
@@ -0,0 +1,193 @@
+    _.containsPattern(TreePattern.UNRESOLVED_TRANSPOSE)) {
+    case t @ UnresolvedTranspose(indexColumnOpt, child) if child.resolved =>
+
+      // Handle empty frame with no column headers
+      if (child.output.isEmpty) {
+        return Transpose(Seq.empty, hasNonIndexColumns = false)
+      }
+
+      // Use the first column as index column if not provided
+      val resolvedIndexColumn = indexColumnOpt match {
+        case Some(indexColumn) if indexColumn.resolved => indexColumn
+        case _ => child.output.head
+      }
+
+      // Cast the index column to StringType
+      val indexColumnAsString = resolvedIndexColumn match {
+        case attr: Attribute if attr.dataType.isInstanceOf[AtomicType] =>
+          Alias(Cast(attr, StringType), attr.name)()
+        case attr: Attribute =>
+          throw new AnalysisException(
+            errorClass = "INVALID_INDEX_COLUMN",
+            messageParameters = Map(
+              "reason" -> s"Index column must be of atomic type, but found: ${attr.dataType}")
+          )
+        case _ =>
+          throw new AnalysisException(
+            errorClass = "INVALID_INDEX_COLUMN",
+            messageParameters = Map(
+              "reason" -> s"Index column must be an atomic attribute")
+          )
+      }
+
+      // Cast non-index columns to the least common type
+      val nonIndexColumnsAttr = child.output.filterNot(
+        _.exprId == resolvedIndexColumn.asInstanceOf[Attribute].exprId)
+      val nonIndexTypes = nonIndexColumnsAttr.map(_.dataType)
+      val commonType = leastCommonType(nonIndexTypes)
+      val nonIndexColumnsAsLCT = nonIndexColumnsAttr.map { attr =>
+        Alias(Cast(attr, commonType), attr.name)()
+      }
+
+      // Exclude nulls and sort index column values, and collect the casted frame
+      val allCastCols = indexColumnAsString +: nonIndexColumnsAsLCT
+      val nonNullChild = Filter(IsNotNull(resolvedIndexColumn), child)
+      val sortedChild = Sort(
+        Seq(SortOrder(resolvedIndexColumn.asInstanceOf[Attribute], Ascending)),
+        global = true,
+        nonNullChild
+      )
+      val projectAllCastCols = Project(allCastCols, sortedChild)
+      val maxValues = sparkSession.sessionState.conf.dataFrameTransposeMaxValues
+      val limit = Literal(maxValues + 1)
+      val limitedProject = Limit(limit, projectAllCastCols)
+      val queryExecution = sparkSession.sessionState.executePlan(limitedProject)
+      val fullCollectedRows = queryExecution.executedPlan.executeCollect()
+
+      if (fullCollectedRows.isEmpty) {
+        // Return a DataFrame with a single column "key" containing non-index column names
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val keyValues = nonIndexColumnsAttr.map(
+          _.name).map(name => UTF8String.fromString(name))
+        val keyRows = keyValues.map(value => InternalRow(value))
+
+        LocalRelation(Seq(keyAttr), keyRows)
```

Comment:
This should also be `Transpose`?

Reply:
Good catch, adjusted, now the spark plan looks like 
```
>>> empty_df.show()
+-------+-------+-------+                                                       
|column1|column2|column3|
+-------+-------+-------+
+-------+-------+-------+
>>> empty_df.transpose().explain(True)
== Parsed Logical Plan ==
'UnresolvedTranspose column1#0: string
+- LogicalRDD [column1#0, column2#1, column3#2], false

== Analyzed Logical Plan ==
key: string
Transpose [key#32], [[column2], [column3]], true

== Optimized Logical Plan ==
LocalRelation [key#32]

== Physical Plan ==
LocalTableScan [key#32]
```